### PR TITLE
feat: add `secretKeyPath` for reading key from fs at build-time

### PR DESCRIPTION
**Pull request by PatrickMi**
Hi 👋
I want to use docker secrets to pass the turnstile secret key for this module. As it looks to me, the turnstile module only accepts the secret key, when it is added to the runtime config by hand. I made a small change to pass the path of a file (in my case the docker secret) containing the secret for turnstile. It would be awesome if you would add that functionality.
I hope this PR can help 💪 Also as I am new to Nuxt I'm not sure if I made some mistakes or could do some things better. So I'm always open for improvement hints.

Greetings from Germany ❤️

### CODE REVIEW

File: playground/nuxt.config.ts
Comment (dargmuesli) [inline playground/nuxt.config.ts:6]:
```suggestion
      // secretKeyPath: 'path/to/secret/key'
```
https://github.com/dargmuesli/dargstack_template is a devops template by me, not related to nuxt :wink:

File: src/development/secrets/turnstile-secret-key.secret.template
Comment (dargmuesli) [inline src/development/secrets/turnstile-secret-key.secret.template:1]: Same, this file is not required in this repository.

File: src/module.ts
Comment (dargmuesli) [inline src/module.ts:8]: Would the recommendation need to be updated? More of a question to Daniel Roe.

File: src/module.ts
Comment (dargmuesli) [inline src/module.ts:13]: Maybe move this directory under the `secretKey` so that the alternative is easy to see.

File: src/module.ts
Comment (dargmuesli) [inline src/module.ts:47]: Wouldn't this be shown every time someone uses a `secretKey` only, without reading from file?

File: src/module.ts
Comment (dargmuesli) [inline src/module.ts:49]: Also, this would always set the `secretKey` to `undefined` if nothing is read from file, I think.

File: src/module.ts
Comment (dargmuesli) [inline src/module.ts:45]:
```suggestion
    const secretKeyFromPath = secretKeyPath && fs.existsSync(secretKeyPath)
        ? fs.readFileSync(secretKeyPath, 'utf-8')
        : undefined
```
(just typed, might need to be changed for the linter)

It's almost always possible to use `const` only. For me a case where I want to use `let` always makes me think twice if I really need it. Mostly style preference only though.

**Review by dargmuesli — COMMENTED** [review]
Also, the README could be updated too, to contain information about the new configuration option.

File: playground/nuxt.config.ts
Comment (PatrickMi) [reply playground/nuxt.config.ts:6]: Well you are right. It just seemed like a good idea to me. I'll remove that comment and the corresponding template file 😁

File: src/module.ts
Comment (PatrickMi) [reply src/module.ts:8]: What do you mean with recommendation as an explanation for me? :)

File: src/module.ts
Comment (PatrickMi) [reply src/module.ts:47]: As far as I can see the setup function is called once when the module is loaded. But maybe I'm wrong.

File: src/module.ts
Comment (PatrickMi) [reply src/module.ts:49]: I wasn't sure how to handle a wrong key input. One solution would be to handle it like a wrong site key, return a warning and 'disable' the turnstile module

File: src/module.ts
Comment (PatrickMi) [reply src/module.ts]: Applied that 👍

File: src/module.ts
Comment (dargmuesli) [reply src/module.ts:8]: > It is recommended you set the secret key via `runtimeConfig.turnstile.secretKey`

File: src/module.ts
Comment (dargmuesli) [reply src/module.ts:47]: Yes, but should it show a warning if a user never intended to make use of secret files?

File: src/module.ts
Comment (dargmuesli) [reply src/module.ts:49]: I mean that using a `secretKey` only as it's currently supported would not work anymore as it appears to me.

File: src/module.ts
Comment (PatrickMi) [reply src/module.ts:47]: I See. I'll change that so the warning will only be shown when the directory of the file for the secret key is passed as an input

File: src/module.ts
Comment (PatrickMi) [reply src/module.ts:49]: When a secret key is passed in the nuxt config then it is used. The secret key from the config has priority over the key from the file

File: src/module.ts
Comment (danielroe) [inline src/module.ts:39]: we should not support setting `secretKeyPath` in the runtime config as this may give a misleading impression that you can set this key at runtime, which will not work - this module only runs at build time

File: src/module.ts
Comment (danielroe) [reply src/module.ts:49]: I think Jonas' point is that you would need to do this instead or you are unsetting `options.secretKey` when it exists:

```suggestion
    options.secretKey = secretKeyFromPath || options.secretKey
```

File: src/module.ts
Comment (PatrickMi) [reply src/module.ts]: Good point! I moved the path from the runtimeConfig to the inlineOptions. Hope this is good now :)

File: src/module.ts
Comment (PatrickMi) [reply src/module.ts:8]: Good point. I also think @danielroe should decide which way to pass the secret key should be the recommended one. If passing it directly via the runtimeConfig remains the recommended way then this does not have to be changed. Also the Readme wouldn't have to be updated to explain how the secret key should be passed

File: README.md
Comment (dargmuesli) [inline README.md:40]: The `turnstile` configuration is already defined. Why add it a second, untyped time under modules?

File: README.md
Comment (dargmuesli) [inline README.md:51]:
```suggestion
  You can pass the secret key either directly in the runtime config or via an environment variable, or you can pass the path to a file that contains the secret key in the module options. This is helpful when using docker secrets where docker passes the secret key via a file.
```

File: src/module.ts
Comment (dargmuesli) [reply src/module.ts:8]: I think this module should recommend usage without dependencies on external tooling, like Docker, and keep the current recommendation. But the README should still document what options are available to use as it now does in this PR :+1:

File: README.md
Comment (dargmuesli) [inline README.md:52]: The secret key is not part of the runtime configuration anymore, so this sentence can be removed I think.

File: src/module.ts
Comment (dargmuesli) [inline src/module.ts:48]: As with other `if`s ;)

```suggestion
      }

      options.secretKey = secretKeyFromPath || options.secretKey
```

**Comment by danielroe** [comment]
I really appreciate the work you both are putting into this.

My own point of view is that the secret should be configured by the user setting an environment variable, rather than reading a file from disk. I'm willing to support reading a file from disk at build-time, however, if you really need it, though I personally can't see why - it seems it would be better to include the following in your `.env` file:

```
NUXT_TURNSTILE_SECRET_KEY=somesecretkey
```

**Comment by dargmuesli** [comment]
Afaik it's not recommended to store secrets in environment variables. It's described in [CWE-526](https://cwe.mitre.org/data/definitions/526.html).
TrendMicro did [research](https://www.trendmicro.com/en_us/research/22/h/analyzing-hidden-danger-of-environment-variables-for-keeping-secrets.html) on that. Diogo Mónica sums it up in a shorter [article](https://blog.diogomonica.com/2017/03/27/why-you-shouldnt-use-env-variables-for-secret-data/). I think these observations are the reason why Docker implemented [secrets](https://docs.docker.com/engine/swarm/secrets/) as files, they write:
> Docker secrets do not set environment variables directly. This was a conscious decision, because environment variables can unintentionally be leaked between containers (for instance, if you use --link).

I see that in this context it might not be very necessary to support secret files, but to me it's better to follow the habit of doing it a safer way.

**Comment by dargmuesli** [comment]
Hey, sorry for the incoming spam, but I am limited by https://github.com/orgs/community/discussions/8327 not being able to resolve my own review comments, so I'm going to comment under those which can be marked as resolved by someone with write access to this PR.

File: playground/nuxt.config.ts
Comment (dargmuesli) [reply playground/nuxt.config.ts]: This thread can be resolved :+1:

File: src/development/secrets/turnstile-secret-key.secret.template
Comment (dargmuesli) [reply src/development/secrets/turnstile-secret-key.secret.template]: This thread can be resolved :+1:

File: src/module.ts
Comment (dargmuesli) [reply src/module.ts:8]: This thread can be resolved :+1:

File: src/module.ts
Comment (dargmuesli) [reply src/module.ts:11]: This thread can be resolved :+1:

File: src/module.ts
Comment (dargmuesli) [reply src/module.ts]: This thread can be resolved :+1:

File: src/module.ts
Comment (dargmuesli) [reply src/module.ts]: I think this change can be reverted as the logic changed how `secretKeyPath` is treated.

File: src/module.ts
Comment (dargmuesli) [reply src/module.ts]: This thread can be resolved :+1:

File: src/module.ts
Comment (dargmuesli) [reply src/module.ts]: As I see the suggested change implemented I'll break the etiquette of letting the thread owner resolve their thread once and mark this conversation as resolved. I hope that saves you some time, Daniel, but feel free to reopen! 🤙

**Review by dargmuesli — APPROVED** [review]

**Pull request merged**
